### ISSUE TITLE
How to build cuda modules in parallel?

### ISSUE THREAD

**Issue by haquocviet** [opencv/opencv#15135]
OpenCV 4.1.0 
Windows 64 Bit
Visual Studio 2019
CUDA Toolkit 10.1

There is no doubt that building cuda modules takes long time. I have realized that the slowness become more severe when cuda source files (*.cu) is compiled one by one under the MSVC. Notes that cpp files are still compiled in parallel as usual. During building cuda source files, I observed the nvcc consumed just a little CPU ~20%.
What I want to know is compiling cuda source files in sequence is the constraint of Nvidia compiler or because of wrong configuration in somewhere. If it's the second case, what is the possible way to force nvcc to build in parallel?

**Comment by j-maestri** [comment]
It looks like you asked this question on the NVIDIA forums as well:
https://devtalk.nvidia.com/default/topic/1057961/does-nvcc-support-compiling-in-parallel-/

According to the NVIDIA moderator Robert Crovella (responded to you on the above forum post), as well as a number of other results online, it seems that there is no easy option for a parallel NVCC build when paired with MSVC without relying on additional tools.  

If you want this functionality and are willing to change compilers/systems, you can try the suggestions listed in this post (also referenced by Robert Crovella, but reposting here for posterity):
https://stackoverflow.com/questions/17657488/cuda-parallel-code-generation-in-visual-studio

This answer is from 2013 and references CUDA 5.5, but from what I can tell still applies to the current CUDA release.

In short: try using the -j flag with make.

**Comment by alalek** [comment]
> no easy option for a parallel NVCC build when paired with MSVC

BTW, building through `ninja` (ninja-build) performs much better than using MSBuild.
